File: 0079-word-search/0079-word-search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */

const dfs = (row, col, word, i, board) => {
    if (i === word.length) return true;
    if (row < 0 || row === board.length || col < 0 || col === board[0].length || board[row][col] !== word[i]) return false;
    const letter = board[row][col];
    board[row][col] = '#';
    const isFound = dfs(row + 1, col, word, i + 1, board) || dfs(row - 1, col, word, i + 1, board) || dfs(row, col + 1, word, i + 1, board) || dfs(row, col - 1, word, i + 1, board);
    board[row][col] = letter;
    return isFound;
}

const exist = (board, word) => {
    const rows = board.length;
    const cols = board[0].length;
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            if (dfs(i, j, word, 0, board)) return true;
        }
    }
    return false;
};

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m = len(board)
        n = len(board[0])
        
        def backtrack(i, j, index, path):
            if index == len(word):
                return True
            if i < 0 or j < 0 or i >= m or j >= n or board[i][j] == '1':
                return False
            if board[i][j] != word[index]:
                return False
            board[i][j] = '1';
            bool = backtrack(i + 1, j, index + 1, path) or backtrack(i - 1, j, index + 1, path) or backtrack(i, j + 1, index + 1, path) or backtrack(i, j - 1, index + 1, path)
            board[i][j] = word[index]
            return bool
            
        for i in range(m):
            for j in range(n):
                if backtrack(i, j, 0, ''):
                    return True
        return False